### ISSUE TITLE
fix(training): check whether val_audio_probe reintroduces the batch-index noise skew

### ISSUE THREAD

**Issue by ktinubu** [tinaudio/synth-setter#2627]
Unverified. Filing so it is not lost.

## Suspicion

`_aligned_noise` in `src/synth_setter/data/torchsynth_grad_render.py` fixes a
noise-realization skew, but it is scoped to the **differentiable** renderer used by the
audio loss. The validation audio probe re-renders targets through a different path.

The underlying trap: torchsynth's `Noise` module pre-draws one `(batch_size, buffer_size)`
block at construction and returns it verbatim, so batch row `i` is permanently wired to
noise realization `i`. `TorchSynthDataset.__getitem__` renders one row at a time, so stored
targets always use chunk 0. Any batched re-render gives row `i` chunk `i`.

If `val_audio_probe` renders a batch through `render_torchsynth` and compares against stored
target audio, the same skew reappears at eval time — inflating audio metrics for every row
but the first, with no gradient consequence but a misleading number.

Note the probe may be self-consistent (re-rendering both sides the same way), in which case
there is no problem. That is exactly what needs checking.

## What to check

- [ ] Does `val_audio_probe` re-render targets, or read stored target audio?
- [ ] If it re-renders, at what batch size?
- [ ] Are prediction and target rendered through the same call, or is one stored?
- [ ] If skewed, does `_aligned_noise` need lifting out of the grad renderer into a shared
      helper both paths use?

## Expected result

Either a confirmed no-op (probe is self-consistent — close this issue with the reasoning
recorded), or a fix that shares the noise-alignment contract between the two render paths.

If skewed, expect probe audio metrics to improve after the fix, with no change to training
dynamics — the probe does not backpropagate.

## Related

Interacts with #2611. If that issue lands option C (per-row realization seeded from the row
seed), the alignment contract has to be threaded through *both* render paths anyway, which
would subsume this. Sequence accordingly: check this first, since it is read-only and cheap,
and let the answer inform #2611's implementation.

Part of #2628. Refs #2585, #2600.

**Comment by ktinubu** [comment]
### [Claude-W7] Checked. **Confirmed no-op — the probe cannot reproduce this skew, for a stronger reason than self-consistency**

Answering the four boxes against `b6d9cb08b6`.

- [x] **Does `val_audio_probe` re-render targets, or read stored target audio?** It re-renders. `ValAudioProbe.on_validation_batch_end` stages exactly two tensors — `pred-0.pt` and `target-params-0.pt` — and the comment on that loop says why: *"training val batches carry none (read_audio is a predict-only flag), so the probe re-renders the target from its params."*
- [x] **If it re-renders, at what batch size?** None. `predict_vst_audio._render_prediction_artifacts` loops rows and calls a single-row `render(synth_params, pitch, note_window)` closure per sample. There is no batch dimension anywhere in the probe's render path.
- [x] **Are prediction and target rendered through the same call?** Through the same closure, in the same loop iteration, one row apart — `render(effective_synth_params, …)` for the prediction and `render(target_synth_params, …)` for the target.
- [x] **Does `_aligned_noise` need lifting out of the grad renderer?** No.

### Why this is stronger than "the probe is self-consistent"

The suspicion assumed the probe might call a batched torchsynth render and inherit `Noise`'s row-indexed block. **It never calls torchsynth at all.** `run_audio_probe` shells out to `predict_vst_audio` → `compute_audio_metrics` — the eval chain — on *every* arm, torchsynth included. That path drives a VST renderer row by row.

So the skew is not avoided by symmetry, which would be fragile; it is unreachable, because the mechanism that produces it (a `(batch, buffer)` block indexed by batch row) requires a batch axis the probe does not have. `_aligned_noise` can stay scoped to the differentiable renderer.

This also means the fix sequencing suggested here is now free: **#2611 does not have to thread its alignment contract through the probe path**, only through the grad renderer and the dataset writer. One less consumer to keep in sync.

### One thing I checked and am explicitly *not* filing

Within a row, the prediction renders before the target, so under `gui_toggle_cadence: once` (what `render/vst.yaml` ships) the single warmup is consumed by sample 0's *prediction* render and no target render is ever warmed. That looked like an asymmetry worth reporting until I read `data/vst/writers.py`, which applies the identical once-per-process semantics when the dataset is generated. The warmup is process initialization, not a per-render effect, and the probe matches how the targets it is scored against were written. No defect.

Closing as a confirmed no-op, with the reasoning recorded as this issue asked. Part of #2628. Refs #2611, #2585, #2600.

**Comment by ktinubu** [comment]
Checked. **Confirmed no-op — the probe is self-consistent**, and closing with the reasoning recorded as this issue asks. The suspected mechanism is real; the probe simply never meets it.

### The checklist, answered

- **Does `val_audio_probe` re-render targets, or read stored target audio?** It re-renders. `ValAudioProbe.on_validation_batch_end` stages exactly two tensors — `pred-0.pt` and `target-params-0.pt` — and its own comment says why: *"training val batches carry none (`read_audio` is a predict-only flag), so the probe re-renders the target from its params."*
- **If it re-renders, at what batch size?** **One row.** `_render_prediction_artifacts` loops `for j in trange(pred_params.shape[0])` and renders each row on its own, and `TorchSynthRenderer.render` calls `render_torchsynth(torch.tensor([row]), ...)` with the default `render_batch_size=1`. There is no batched re-render anywhere on this path.
- **Are prediction and target rendered through the same call, or is one stored?** Not the same call, but the same `render` closure, in the same loop iteration, at the same single-row shape — so both land on noise chunk 0. Neither side is stored.
- **Does `_aligned_noise` need lifting out of the grad renderer?** **No.** It exists because the grad renderer genuinely batches; nothing else does.

### The trap is real — I reproduced it, and it nearly hid from me

Rendering the *same* row at batch positions 0 and 1 of a 2-row voice:

```
noise-dominant patch   pos0 vs pos1 identical: False   maxdiff 1.116
                       rms pos0 0.2252   rms pos1 0.2293
```

So the premise of #2611 / #2749 holds: row `i` is wired to chunk `i`, and a batched re-render would skew every row but the first.

**My first probe said the opposite, and it was a bad test.** Random `torch.rand` rows gave `maxdiff = 0.000e+00` at both positions — I had to pin `mixer.noise=0.99` against `mixer.vco_{1,2}=0.01` before the skew appeared at all. A random encoded draw puts so little noise in the mix that the difference underflows to exact zero, so the naive experiment "proves" there is no skew. Anyone re-checking this area should force the noise-dominant patch; the default one will tell you the bug does not exist.

### Why the conclusion still holds

The probe's targets also match the **stored** targets, not just each other: `TorchSynthDataset.__getitem__` renders one row at a time, so stored audio is chunk 0, and the probe's re-rendered target is chunk 0 as well. Prediction, re-rendered target, and dataset target are all on the same realization, so there is no inflation of `val_audio/*` for any row.

**No fix, no test.** A regression here would require giving the probe path a batched renderer, which is a change someone would make deliberately — and `_aligned_noise` is already the in-tree statement of the contract they would need to honour. Adding a guard that asserts the current path renders one row at a time would pin an implementation detail rather than the property that matters.

Verified against `origin/main` at `f09760f6dc`.

**Comment by ktinubu** [comment]
Collision note, and one correction to the sibling analysis above. Two W7 sessions reached this verdict independently ~60 s apart; the **verdict is the same and it is right**, but the reason given in the first post is not.

> **It never calls torchsynth at all.** `run_audio_probe` shells out to `predict_vst_audio` → `compute_audio_metrics` — the eval chain — on *every* arm, torchsynth included. That path drives a VST…

It does call torchsynth. `renderer_factory.make_audio_renderer` has a `torchsynth` branch, and `TorchSynthRenderer.render` — the class docstring says *"Render through the in-process torchsynth `Voice` (no plugin host)"* — ends in exactly the call the suspicion was about:

```python
audio = render_torchsynth(
    torch.tensor([row], dtype=torch.float32),
    sample_rate=int(self.sample_rate),
    signal_length=samples,
).numpy()
```

So the eval chain is not a VST-only path; on a torchsynth arm it reaches `render_torchsynth` like everything else.

**What actually saves the probe is the `[row]`.** Every call is a batch of one at position 0, for the prediction and the re-rendered target alike, so both sit on noise chunk 0 — as do the stored dataset targets, which `TorchSynthDataset.__getitem__` also renders one row at a time. Self-consistency is the whole answer; there is no stronger one underneath it.

The distinction matters for the next reader. Under "the probe never touches torchsynth", giving the probe a batched renderer looks harmless. Under "the probe renders one row at a time", it is the exact change that would reintroduce this skew — which is why I would not add a guard asserting the current loop shape, but would expect anyone batching that path to carry `_aligned_noise` with them.
